pkg/crc/image/image.go

```diff
@@ -0,0 +1,112 @@
+package image
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"runtime"
+	"strings"
+
+	crcpreset "github.com/code-ready/crc/pkg/crc/preset"
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/transports/alltransports"
+	"github.com/containers/image/v5/types"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+var (
+	registryURI = "quay.io"
+	orgName     = "crcont"
+)
+
+type OCIImageHandler struct {
+	ImageName string
+	ImageTag  string
+	Progress  io.Writer
+}
+
+func (img *OCIImageHandler) uri() string {
+	return strings.Join([]string{registryURI, orgName, img.ImageName}, "/")
+}
+
+func (img *OCIImageHandler) parse(image string) (types.ImageReference, error) {
+	return alltransports.ParseImageName(image)
+}
+
+func (img *OCIImageHandler) policyContext() (*signature.PolicyContext, error) {
+	policy := &signature.Policy{Default: []signature.PolicyRequirement{signature.NewPRInsecureAcceptAnything()}}
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return nil, fmt.Errorf("error creating security context: %s", err.Error())
+	}
+
+	return policyContext, nil
+}
+
+func (img *OCIImageHandler) getSystemContext() *types.SystemContext {
+	return &types.SystemContext{
+		OSChoice: "linux",
+	}
+}
+
+// CopyImage pull the image from the registry and put it to local storage
+func (img *OCIImageHandler) CopyImage(destPath string) ([]byte, error) {
+	// Source Image from docker transport
+	srcImg := fmt.Sprintf("docker://%s:%s", img.uri(), img.ImageTag)
+	srcRef, err := img.parse(srcImg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid source image name %s: %v", srcImg, err)
+	}
+
+	destImg := fmt.Sprintf("dir:%s", destPath)
+	destRef, err := img.parse(destImg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid destination name %s: %v", destImg, err)
+	}
+
+	policyContext, err := img.policyContext()
+	if err != nil {
+		return nil, err
+	}
+
+	return copy.Image(context.Background(), policyContext,
+		destRef, srcRef, &copy.Options{
+			ReportWriter: img.Progress,
+			SourceCtx:    img.getSystemContext(),
+		})
+}
+
+func GetBundleImageLayerFromManifest(manifestData []byte) (string, error) {
+	manifest := &v1.Manifest{}
+	// Get the file layer sha from manifest
+	if err := json.Unmarshal(manifestData, manifest); err != nil {
+		return "", err
+	}
+	return removeShaFromDigestString(manifest.Layers[0].Digest.String()), nil
```

Comment:
not sure this deserves a separate helper since it's going to be a one-liner used once, and this method is also not very long.
I tried this code with multi-layer images, and it did not work at all (expected, and not saying it has to be fixed). I was wondering if there's a way of detecting this though, maybe disallow multi-layer images, and also check the `MediaType` to make sure it's the expected one?

Reply:
> I tried this code with multi-layer images, and it did not work at all (expected, and not saying it has to be fixed). I was wondering if there's a way of detecting this though, maybe disallow multi-layer images, and also check the `MediaType` to make sure it's the expected one?

I am not sure why it shouldn't work with multi layer, the logic here is just return the first layer digest, in your test does the first layer didn't have any digest?

Reply:
It will do "something" for multilayer images, but will it be doing the right thing, or something unexpected to the user? Imo it's not very well defined what will happen, so it can just error out so that we catch immediatly that something was off.